### ISSUE TITLE
Consider Kibana resource watcher

### ISSUE THREAD

**Issue by anyasabo** [elastic/cloud-on-k8s#2155]
We should consider creating a watcher tool for Kibana, that can watch config maps (or some other type of resource) and automatically import them into Kibana. This would let users have the same set of Kibana resources across environments, and allow for easy distribution (say, we could supply some default dashboards to observe the operator's metrics). My thinking is to initially support Kibana dashboards, but potentially expand it to Watcher watches. There may be more resources.

It's not obvious to me where this would live -- I'm leaning towards the operator, though it may make sense for it to be a sidecar in Kibana too. I think the downside is that you're running a container for replica when we really only need one per Kibana.

### Notes

The kibana team recommended we use the saved objects import/export API:
https://www.elastic.co/guide/en/kibana/current/saved-objects-api-import.html

As the import/export dashboard API is being removed in 8.0:
https://github.com/elastic/kibana/issues/41439

### Examples in other projects
Grafana supports declaring them in config files that are loaded on startup:
https://grafana.com/docs/administration/provisioning/#dashboards

which is how kube-prometheus project does it:
https://github.com/coreos/kube-prometheus/blob/master/manifests/grafana-deployment.yaml#L42

https://github.com/coreos/kube-prometheus/blob/master/docs/developing-prometheus-rules-and-grafana-dashboards.md#dashboards

though there are projects using techniques like described here (and what kube-prom used to do) with a watcher for config maps:
https://github.com/PierreVincent/k8s-grafana-watcher

**Comment by mtparet** [comment]
It looks like CRDs ! :)

**Comment by jbnjohnathan** [comment]
To include a KibanaDashboard or similar CRD would be very nice. Is there any progress to mention on this ticket yet?

**Comment by captainpro-eng** [comment]
+1

**Comment by AlanMasciangelo** [comment]
Is there an alternative recommended approach for managing saved objects?

**Comment by Stevenpc3** [comment]
We currently deploy the ECK operator and stack using helm charts.  We deploy logstash that creates our indices but then we have to go and manually import our objects under stack management like our dashboards.  We manually create our data views but they could be objects too.  We considered how it would be good to load these via configmaps.  We at one point loaded the objects using kube lifecyclehooks in the pod template via command line api calls, but this will not work because of the chicken-and-egg problem of data-view needs an index and dashboards needs index and data-views.  This poses a problem because the index is not created until logstash starts which we cannot guarantee kibana will wait for logstash.  So it would be nice to have logstash load objects into kibana as a final step.

To get around this we also had our services that need a dashboard do the heavy work.  The service will start and wait on elasticsearch then it will use the api to create an index in elasticsearch then after that it uses the kibana api to load a dashboard into elasticsearch.  The fact that the dashboard is stored in elasticsearch but can only be loaded via kibana is part of the issue.

Having CRDs or configmaps or secrets that can load the objects in and ensuring that kibana dosen't error out if the dataview, dashboard, and index are loaded out of order would be required.

The goal is for us to be able to deploy via helm and have all of our dashboards and other objects included in the deployment so it is repeatable and consistent without manual steps like we have to do now... or hacks.
